### ISSUE TITLE
Changing zElevation often creates large performance impact compared to changing x or y position.

### ISSUE THREAD

**Issue by MikalDev** [Scirra/Construct-bugs#7604]
When changing zElevation often it creates a large CPU performance impact compared to changing x or y position. In my case the reason z may change as often as x or y is I am using a 3d physics model to develop games. In this case x,y,z may all change as physics is processed. The impact appears on both setting zElevation in script and increased drawCall performance impact.

## Attach a .c3p
[ZElevationTest.zip](https://github.com/Scirra/Construct-bugs/files/13537243/ZElevationTest.zip)

## Steps to reproduce

1. Change zip tp c3p
2. Run project
3. Click x and z buttons and compare CPU usage (note that in the x case the z elevation is fixed, but each shape can have a different value.)

## Observed result

CPU usage is much higher for changing zElevation compared to changing x position.

## Expected result

CPU usage for z elevation change and x change to be similar.

## More details

Looking at chrome perf and source code, I see StateReleaseGroup and new batches being formed as z elevation changes, which may account for the additional performance requirement.

Chrome

**Affected browsers/platforms:** 

**First affected release:** 

## System details

PASTE HERE
<img width="675" alt="Screenshot_2023-12-02_at_9 27 24_AM" src="https://github.com/Scirra/Construct-bugs/assets/33675273/13fc80f7-781d-4f67-bee1-249790dee9f0">
<img width="675" alt="Screenshot_2023-12-02_at_9 22 39_AM" src="https://github.com/Scirra/Construct-bugs/assets/33675273/d30ab306-794e-4b7f-b2a7-85bcce3f18e8">
<img width="675" alt="Screenshot_2023-12-02_at_9 22 49_AM" src="https://github.com/Scirra/Construct-bugs/assets/33675273/a3c48f96-7e18-42b7-848d-d45599dcf9ce">
<details><summary>View details</summary>
</details>

**Comment by MikalDev** [comment]
In my case, the focus is on 3D performance, so if there are requirements such as only 3D Shape on a particular will activate a faster path, that would also work for me (w/ a fallback in case other objects are placed on the level.)

**Comment by AshleyScirra** [comment]
Changing Z elevation is fundamentally more CPU intensive than changing the X position, because changing Z elevation is a change to the Z order, which changes the draw order, which then requires a sort. So the fact changing Z elevation is more CPU intensive is expected.

I tried tweaking some of the code around this but it didn't make that much difference, and changes would probably impact the general performance of 2D games as well, so it's a classic tradeoff of what we want to optimise, and Construct remains primarily a 2D engine. In my view the difference is also relatively small for a project with 3000 objects, so I don't think it's that bad a problem - the performance seems OK both ways as it is. But even if this was better optimised, it would still be slower changing the Z elevation than X position because of the draw order sorting, so that aspect of it is not feasible to solve.

**Comment by MikalDev** [comment]
Thanks for the feedback and taking a look at it. I understand on the compromise. In some ways I want a full 3d mode that doesn't have layers or draw order sorting (though transparency....). This probably is more of a Construct 3D version than a Construct 3 version though and I understand your priorities.

Now that you do know the request, I hope it goes in your mind as a little seed and unexpectedly blooms into a brilliant solution in the fertile field of your subconscious one day :)

**Comment by F3der1co** [comment]
@AshleyScirra 
The performance hit still remains even when setting the draw order to camera distance, with that setting a draw order sort wouldn't be necessary with z-elevation changes from my understanding, so couldn't it be skipped in that specific case? This would also only affect 3D games as this sorting order is only available for 3d, this could be a good solution as it wouldn't negatively affect 2d projects. I hope you can take another look.

**Comment by MikalDev** [comment]
Good point F3der1co. Also consider the case where the camera is parallel to the layout. In this case, zElevation sort does not indicate render order, which is instead based on x,y distance from camera.
